File: addons/mail/static/tests/mock_server/mock_models/discuss_channel_category.js
import { fields, getKwArgs, makeKwArgs, models } from "@web/../tests/web_test_helpers";
import { DiscussChannel } from "./discuss_channel";
import { mailDataHelpers } from "../mail_mock_server";

export class DiscussChannelCategory extends models.ServerModel {
    _name = "discuss.channel.category";

    name = fields.Char({ string: "Channel category name", required: true });
    channel_ids = fields.One2many({
        relation: "discuss.channel",
        relation_field: "channel_category_id",
        string: "Channels",
    });

    get_channels(ids) {
        const kwargs = getKwArgs(arguments, "ids");
        ids = kwargs.ids;
        delete kwargs.ids;

        const [channel_category] = this.browse(ids);
        return new mailDataHelpers.Store(
            DiscussChannel.browse(channel_category.channel_ids)
        ).get_result();
    }

    /** @param {number[]} ids */
    _to_store(ids, store) {
        const kwargs = getKwArgs(arguments, "ids", "store", "fields", "extra_fields");
        ids = kwargs.ids;

        const fields = {
            name: true,
        };

        const categories = this.browse(ids);
        for (const category of categories) {
            const [data] = this.read(category.id, Object.keys(fields), makeKwArgs({ load: false }));
            store.add(this.browse(category.id), data);
        }
    }
}

File: addons/mail/models/discuss/discuss_channel_category.py
# Part of Odoo. See LICENSE file for full copyright and licensing details.

from odoo import fields, models
from odoo.addons.mail.tools.discuss import Store


class ChannelCategory(models.Model):
    _name = "discuss.channel.category"
    _description = "Channel Category"

    name = fields.Char("Channel category name", required=True)
    channel_ids = fields.One2many("discuss.channel", "channel_category_id", string="Channels")
    sequence = fields.Integer(string='Sequence', default=5)

    def get_channels(self):
        self.ensure_one()
        return Store(self.channel_ids).get_result()

    def _to_store(self, store: Store):
        """Adds channel categories data to the given store."""
        fields = {"name": True, "sequence": True}
        for category in self:
            data = category._read_format(list(fields), load=False)[0]
            store.add(category, data)
